docs/rfc/rfc-9.adoc

```diff
@@ -0,0 +1,145 @@
+optimistically mint using the same Minter & Guardian process we already have. In
+the background, we sweep to disable the refund. Later, someone brings back the
+NFT and a <<Fees,fee>>. We burn the NFT and send them their Ordinal.
+
+=== Depositing Details
+
+Depositing works essentially the same way that it does in tBTCv2 with a few key
+differences.
+
+Once a deposit is revealed, it is **extremely** important that the wallet does
+not sweep them into the same UTXO; that would be destructive. We still need to
+sweep to disable the refundfootnote:[Whether or not the sweep can be performed
+in a batch is a further research item.]. The sweep needs to preserve the
+individual satoshis.
+
+Once custodied, we would mint an ERC721 where the `metadata` field references
+the https://docs.ordinals.com/overview.html[Ordinal ID]. They offer multiple
+notations; using the Integer Notation looks simple. The reference to the Ordinal
+ID is enough for a dApp to pull the underlying content, but it needs to be
+plugged in to both Bitcoin and Ethereum.
+
+The ERC721 in the user's Ethereum Wallet points at the Ordinal ID. Then, the
+https://github.com/casey/ord[ord] program can identify the associated Ordinal on
+BTC, and extract its payload, which looks like:
+
+```
+OP_FALSE
+OP_IF
+  OP_PUSH "ord"
+  OP_1
+  OP_PUSH "<content-type>" // example: OP_PUSH "text/plain;charset=utf-8"
+  OP_0
+  OP_PUSH "<content>" // example: OP_PUSH "Hello, world!"
+OP_ENDIF
+```
+
+=== Redemption Details
+
+Since NFTs are unique, there is no longer choice about which wallet a NFT is
+redeemed from. It **must** be the wallet that is the custody of that Ordinal.
+
+The redemption process is otherwise identical to tBTCv2. Someone sends the NFT,
+a <<Fees,fee>>, and an ethereum address back to contracts. The associated wallet
+sends them their Ordinal to the provided address with a 51-of-100 signature.
+
+=== Fraud Detection
+
+Rather than storing the full list of custodied Ordinals per
+walletfootnote:[Storing the full list would of Ordinals be very gas
+inefficient.], we instead:
+
+* Take the merkle root of all Ordinals under custody
+* Propose this root on chain
+* Let anyone challenge this for a reward if it is wrong
+
+This means we are efficiently sure that we have the right merkle root and
+underlying custody.
+
+Later, there are two sorts of misbehavior from the custody side:
+
+(1) Sending an Ordinal to an address that did not redeem it and not having it be
+  associated with a wallet closure.
+
+(2) Not redeeming an Ordinal within the allotted time.
+
+Anyone can challenge any time (at a cost) for (1). The challenge is defeated by
+providing the legimate request within the timeout.
+
+In order to create challenge (2), the challenger must provide a proof that the
+Ordinal is being custodied and that a redemption was requested. The challenge is
+defeated by providing a SPV proof that the Ordinal was properly transferred.
```

Comment:
Are these easy to verify? Considering we need per-satoshi tracking.

Reply:
Oh sheesh, that's a good question; we can prove that a transaction happened in a block with SPV proofs and our relay, but i'm not sure yet if we can actually prove anything about it being the Ordinal that it says it is. Further research!

Reply:
> Considering we need per-satoshi tracking.

This looks like a problem with proving the ordinal is custodied after sweeping. This could be done with SPV proof on the sweep transaction and per-ordinal tracking which feels like a one-write to a mapping on EVM. We can then construct redemption challenges.

Reply:
[Here is what we know](https://github.com/keep-network/tbtc-v2/blob/5d6232bb50660ae04f5c4af3fc2f53bc989ab251/solidity/contracts/bridge/BitcoinTx.sol#L24-L92) on Ethereum about Bitcoin transaction that is proven. 

I also realized I do not fully understand inscriptions. 

Each sat is unique - it is numbered. So we can say each sat is a kind of an NFT but with no metadata attached. Those single sats are... ordinals? 

Now, having uniquely numbered tokens is not enough. They need to contain some metadata. Ordinal + metadata = inscription? 

If I understand it, correctly, the metadata is stored in
```
OP_FALSE
OP_IF
  OP_PUSH "ord"
  OP_1
  OP_PUSH "<content-type>" // example: OP_PUSH "text/plain;charset=utf-8"
  OP_0
  OP_PUSH "<content>" // example: OP_PUSH "Hello, world!"
OP_ENDIF
```

this looks like a standard Bitcoin script in an input. Is it included only in the "original input" when the inscription was created and every other transfer is based on the sat number (ordinal)? 

If so - if extracting sat number from the [transaction data](https://github.com/keep-network/tbtc-v2/blob/5d6232bb50660ae04f5c4af3fc2f53bc989ab251/solidity/contracts/bridge/BitcoinTx.sol#L24-L92) is doable (I do not know if, did not validate), we can prove on Ethereum the given ordinal was deposited but we will not know the metadata. Going back 100 transactions may not be feasible given EVM limitations. We need to come up with some other algorithm for that.

I do not think we need to have it all in this RFC, but I think we should mark the spots that require future research. My understanding - after reading the RFC for the first time - was that just a few slight changes are needed but I am not really sure about it 🤔

Reply:
> Each sat is unique - it is numbered. So we can say each sat is a kind of an NFT but with no metadata attached. Those single sats are... ordinals?

> Now, having uniquely numbered tokens is not enough. They need to contain some metadata. Ordinal + metadata = inscription?

Yup! All correct including how the metadata is stored

> Is it included only in the "original input" when the inscription was created and every other transfer is based on the sat number (ordinal)?

That's what I keep looking for; I'm having a bunch of trouble finding an ordinal explorer that links me to the underlying satoshi and where it's being kept. I might just need to send one I have around and post logs.

> If so - if extracting sat number from the [transaction data](https://github.com/keep-network/tbtc-v2/blob/5d6232bb50660ae04f5c4af3fc2f53bc989ab251/solidity/contracts/bridge/BitcoinTx.sol#L24-L92) is doable (I do not know if, did not validate), we can prove on Ethereum the given ordinal was deposited but we will not know the metadata.

**We** should never need to know the inscription metadata (or even that it has some); we only care about the Ordinal ID. For example, say you mint Ordinal #12345 and it has metadata which is a PNG of a cat. You send that to our bridge. We make a NFT with metadata "Ordinal #12345".

Then you hook up your wallet to an ethereum dapp and it says "your NFT has metadata pointing to Ordinal #12345. Ordinal #12345 points to a PNG of a cat. This is what it looks like".

The dapp would have to be aware of our NFT reference format and be able to pull the metadata from bitcoin, but that's relatively straightforward.

That said, I have no idea if we can extract the sat number from the tx data